stream/stream_test.go

```diff
@@ -0,0 +1,139 @@
+package stream_test
+
+import (
+	"context"
+	"errors"
+	"iter"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/creachadair/chirp"
+	"github.com/creachadair/chirp/peers"
+	"github.com/creachadair/chirp/stream"
+)
+
+func TestStream(t *testing.T) {
+	tests := []struct {
+		in      string
+		want    []string
+		wantErr string
+	}{
+		{"stream foo bar", vals("foo", "bar"), ""},
+		{"stream foo bar, err", vals("foo", "bar"), "service error: test"},
+		{"err", vals(), "service error: test"},
+		{"req, req, stream foo", vals("req", "req", "foo"), ""},
+		// server-side cancellation just before successful stream end
+		{"stream foo, server-cancel", vals("foo"), "service error: context canceled"},
+		// server-side cancellation that HandlerFunc ignores
+		{"stream foo, server-cancel, stream bar qux", vals("foo"), "service error: context canceled"},
+		// server-side cancellation that HandlerFunc obeys
+		{"stream foo, server-cancel, return-canceled", vals("foo"), "service error: context canceled"},
+		// client-side cancellation
+		{"stream foo, client-cancel, stream bar qux", vals("foo"), "context canceled"},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.in, func(t *testing.T) {
+			ps := peers.NewLocal()
+
+			ctx, clientCancel := context.WithCancel(context.Background())
+			defer clientCancel()
+
+			stream.Handle(ps.B, "stream", parseStreamSpec(t, tc.in))
+			ps.B.NewContext(func() context.Context {
+				// Give the server-side handler access to both client-side
+				// and server-side CancelFuncs, so that parseStreamSpec
+				// can drive cancellation on either end of the Peer. To
+				// synchronize client-side cancellation properly, also
+				// smuggle in the client-side context, which will be used
+				// exclusively for synchronizing cancellation.
+				serverCtx, serverCancel := context.WithCancel(context.Background())
+				serverCtx = context.WithValue(serverCtx, serverCancelContextKey{}, serverCancel)
+				serverCtx = context.WithValue(serverCtx, clientCtxContextKey{}, ctx)
+				serverCtx = context.WithValue(serverCtx, clientCancelContextKey{}, clientCancel)
+				return serverCtx
+			})
+
+			var got []string
+			var gotErr error
+			for resp, err := range stream.Call(ctx, ps.A, "stream", []byte("req")) {
+				if err != nil {
+					gotErr = err
+					break
+				}
+				got = append(got, string(resp))
+			}
+			if !reflect.DeepEqual(got, tc.want) {
```

Comment:
I usually like to use `cmp.Diff` for these, FWIW, since it makes nicer output when the test fails (e.g., https://github.com/creachadair/chirp/blob/main/chirp_test.go#L118)